bootstrap/kubeadm/internal/ignition/clc/clc.go

```diff
@@ -0,0 +1,376 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package clc generates bootstrap data in Ignition format using Container Linux Config Transpiler.
+//
+// CLC configuration defined in this package will run kubeadm command by creating a /etc/kubeadm.sh script
+// file containing both pre and post kubeadm commands as well as the kubeadm command itself.
+//
+// /etc/kubeadm.sh script will be executed using kubeadm.service systemd unit, which will only happen
+// if /etc/kubeadm.yml file exists, which ensures the script will run only once, as by the end of the
+// script, /etc/kubeadm.yml is moved to /tmp filesystem, so it gets automatically cleaned up after a
+// reboot. This is to align the implementation with cloud-init, which places kubeadm configuration in
+// /tmp directory directly, which is not possible with Ignition.
+//
+// /etc/kubeadm.yml file contains generated kubeadm configuration and can be customized using pre kubeadm
+// commands if needed, as a replacement for Jinja templates supported by cloud-init, for example
+// using 'envsubst' or 'sed'.
+//
+// To override the behavior of kubeadm.service unit, one should create a override drop-in for this it
+// using AdditionalConfig field. Data from this field takes precedence and will be merged with
+// configuration generated by the bootstrap provider, overriding already defined fields following the
+// merge strategy described in https://coreos.github.io/ignition/operator-notes/#config-merging.
+package clc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/template"
+
+	ignition "github.com/coreos/ignition/config/v2_3"
+	ignitionTypes "github.com/coreos/ignition/config/v2_3/types"
+	clct "github.com/kinvolk/container-linux-config-transpiler/config"
+	"github.com/pkg/errors"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/cloudinit"
+)
+
+const (
+	clcTemplate = `---
+{{- if .Users }}
+passwd:
+  users:
+    {{- range .Users }}
+    - name: {{ .Name }}
+      {{- if .Gecos }}
+      gecos: {{ .Gecos }}
+      {{- end }}
```

Comment:
We could use `with` to avoid duplicating accessed keys.
```suggestion
      {{- with .Gecos }}
      gecos: {{ . }}
      {{- end }}
```